sogs/events.py

```diff
@@ -0,0 +1,158 @@
+    if not event_name_valid(name):
+        raise Exception('invalid event name: {}'.format(name))
+    global _pools
+    pool = _pools[name]
+    if msg.conn in pool:
+        raise Exception('already subscribed to {}'.format(name))
+    pool.append(conn)
+
+
+def _handle_subscribe_request(msg):
+    """ handle a request to subscribe to events as a bot"""
+    for name in msg.dataviews():
+        name = name.decode('utf-8')
+        if name in ('all', '*'):
+            # subscribe to all events
+            for ev in EVENTS:
+                _sub_conn_to_event(ev, msg.conn)
+            return
+        _sub_conn_to_event(name, msg.conn)
+    else:
+        raise Exception("no events selected")
+
+
+def _handle_find_room(metric, query, encode=json.dumps):
+    """ find a room """
+    if metric == 'id':
+        room = model.Room(id=int(query))
+    elif metric == 'token':
+        room = model.Room(token=query)
+    else:
+        raise Exception("invalid query metric: {}".format(metric))
+    reply = {a: getattr(room, a) for a in model.Room.ALL_PROPS}
+    reply['mods'] = room.get_mods()
+    reply['active_users'] = room.active_users()
+    return encode(reply)
+
+
+def _handle_find_user(metric, query, encode=json.dumps):
+    """ find a user """
+    if metric == 'id':
+        user = model.User(id=int(query))
+    elif metric == 'pubkey':
+        user = model.User(session_id=query)
+    else:
+        raise Exception("invalid query metric: {}".format(metric))
+    reply = {a: getattr(user, a) for a in model.User.ALL_PROPS}
+    return encode(reply)
+
+
+def _handle_find_request(msg):
+    """ finds a user / room by id / token / pubkey """
+    parts = msg.dataviews()
+    if len(parts) != 3:
+        raise Exception('3 arguments required: entity-kind, query-metric, query-value')
+
+    kind = parts[0].decode('utf-8')
+    metric = parts[1].decode('utf-8')
+    query = parts[2].decode('utf-8')
+
+    _kinds = {'room': _handle_find_room, 'user': _handle_find_user}
+
+    if kind in _kinds:
+        return _kinds[kind](metric, query)
+    raise Exception("cannot find a '{}' we dont have those".format(kind))
+
+
+def _handle_mod_ban(msg):
+    """ handle a ban user from room rqeuest """
+    parts = msg.dataviews()
+    if len(parts) < 2:
+        raise Exception("Not enough arguments, need 2 arguments: user_id, room_id")
+    room_id = int(parts[1].decode('utf-8'))
+    user_id = int(parts[0].decode('utf-8'))
+    room = model.Room(id=room_id)
+    ban_user = model.User(id=user_id)
+    bot = _user_from_conn(msg.conn)
+    if not model.ban_user(bot, room, ban_user):
+        raise Exception("user not banned")
+
+
+def _propagate_event(eventname, *args):
+    """ propagate an event to everyone who cares about it """
+    assert event_name_valid(eventname)
+    global _mq, _pools
+    assert _mq is not None
+    for conn in _pools[eventname]:
+        _mq.command('sogs.event', conn, eventname, *args)
```

Comment:
This would be a bit more convenient if we did `*(bt_serialize(a) for a in args)` so that the caller can pass in a dict/list/whatever.